### DISCUSSION TITLE
Lag/buffering with experimental LA build (E3v2)

### DISCUSSION THREAD

**Discussion by preyx** [mriscoc/Ender3V2S1#386] [General]
(Fully aware I'm using an experimental build and YMMV applies, but would like to give some feedback to help for future dev and possibly find a solution!)

I've always been curious about trying linear advance. I know it's not optimized for bowden setups, but it's still gotten me results! Greatly decreased corner blobbing, and it even allowed me to reduce my retractions to 2mm!

Ender 3 v2, purchased spring 2022, 4.2.2 board:
- Using "Experimental-Ender3V2-422-BLTUBL-LA-20220817.bin" as downloaded, no alterations.
- CR Touch installed.
- Factory extruder + aluminum parts upgrade.
- Factory mk8 end + BTT heatbreak.
- Factory bowden layout, no Capricorn.
- Using linear advance = 0.420, and junction deviation = 0.010.
- Using UBL and 3-point (G29) start code. Z-fade = 2.0mm
- 0.4mm nozzle, 0.2mm layer height.
- Slicing with PrusaSlicer 2.5.0 final.

I've run a dozen or so prints since installing this firmware. A few, but not all, of my prints have an issue during the first 1-2mm where the hotend stutters on some curves, resembling buffering issues. It ends up causing blobbing in the affected areas. I observed that the UI also becomes unresponsive at these times, presumably due to CPU maxing out. Interestingly enough, the issue doesn't affect all curves on the layer, and it isn't persistent throughout the entire print. It also doesn't appear to lag when making the same curve later on (through-holes, for example). This issue has shown up with varying models and prints, but appears to pop up more often with increasing complexity.

I'm unsure if the slicer or g-code has an impact on this, so I have attached a gcode ([sample-LA-gcode.txt](https://github.com/mriscoc/Ender3V2S1/files/9593372/sample-LA-gcode.txt)) that caused the issue for me as an example. I don't have a direct-drive version of my printer to compare with, so I'm also unsure if the aggressive LA value may be related as well.

Other than that, I've been having a great experience wth this firmware so far! The July 2022 build was the first custom firmware I tried on this printer and I've never looked back. Thanks again.

**Comment by mriscoc** [comment]
Hi, thank you for your report. Are you printing from SD or Host (Octoprint)? Did you try with another SD card?

**Reply by preyx** [reply]
Forgot that important detail! I'm using SD cards, no modifications to the printer card slot. I just tested with another SD card I use with this printer. Similar results, stuttering in the first 1 mm or so, then seems to complete at full speed without issue.

In the gcode I shared above, the stuttering is most noticable around the circle feature on the low-X end of the print. For example, around lines 1434-1458 and 1507-1542.

**Reply by mriscoc** [reply]
Try with power outage recovery disabled. Some SD card brands have a very slow writing speed at the protocol that the board uses.

**Reply by preyx** [reply]
Holy cow you do know your stuff! Yes I did have power outage recovery on. I turned it off, ran the same print, and this time it appears to have no stuttering or blobbing. Didn't realize that feature could induce that much lag.

**Reply by DanMan32** [reply]
I also was going to suggest disabling the outage recovery.  I watched a YT about this where curves can be a problem because in reality, curves are really lots of small straight lines and that can fill the print buffer quickly, especially when the printer has to pause reading from the SD card so the printer can write a checkpoint to the SD card.  Using LA may compound the problem with more data to process.
Besides turning off outage recovery, I think the YT had another workaround solution.
But many have had too many issues with resumes that it might not be worth it.  I've tried it when I intentionally paused a print then powered off since I was going to be out of the house for a while.   The part that had originally printed shrunk and warped, and the print started from the beginning anyway, even though I told it to resume.
Better to get a UPS to at least allow for brief power interruptions than rely on power outage resume.

**Comment by curather** [comment]
I have been testing the LA .  I have not had this buffering issue.  Here is my setup

- Using "Experimental-Ender3V2-422-BLTUBL-LA-20220817.bin" as downloaded, no alterations.
- CR Touch installed.
- Direct drive factory with aluminum part upgrade
- Factory mk8 end + BTT heatbreak.
- Herome 6 setup with dual 5015 part cooler and 4020 cooler fan
- Using linear advance = 0.260, and junction deviation = 0.010.
- Using UBL and Z-fade = 2.0mm
- 0.4mm nozzle, 0.2mm layer height.
- Slicing with Cura


I did do a Linear advance test and noticed that if I set the linear advance too high it would cause the blobbing and break the line between the fast and slow side.  I think the op might be having this issue.

**Reply by mriscoc** [reply]
Thank you for your confirmation.

**Reply by curather** [reply]
Glad I could help.

I tested this out with octoprint and notice that there is some type of buffering going on sometime.  I think it has to due with the slow sending of individual lines if you are printing from the pi hard drive.  This buffering that happens from octoprint from what I noticed was random and lasted a few seconds sometimes.  This was with a 40mb gcode file on the hard drive.

**Comment by DanMan32** [comment]
Question: do you have a 2208 stepper driver on your 4.2.2 or one that is supposed to be more compatible with LA?
I am fairly certain I have a 2208 based on T8 printed on a sticker over the card reader, and that's supposed to mean I have a 2208.
I'm thinking of trying LA myself.

**Reply by preyx** [reply]
I have a v2. I checked online and all E3v2 4.2.2 boards should come with 2208 from the factory. So far so good with LA (and power recovery off), recent PLA tolerance test is good to 0.15mm, which is more than enough for all my projects. Highly recommend if you have the supporting hardware.

**Reply by DanMan32** [reply]
There are at least 4 variants to the 4.2.2 board.  A marking or sticker on the SD card reader indicates which stepper driver variant you have.  Up until now, the variant with the 2208 was not capable of LA.
E = A4988
A = TMC2208
B = TMC2209
H = TMC2225

I believe T8 is also TMC2208, and I have a T8 sticker over my SD card reader.

**Reply by mriscoc** [reply]
> the variant with the 2208 was not capable of LA.

Boards with 2208 can have LA: https://github.com/mriscoc/Special_Configurations/releases/tag/LinearAdvance

**Reply by DanMan32** [reply]
I wish that article explicitly addresses the 2208 issue and describes how it was solved.

**Reply by mriscoc** [reply]
> I wish that article explicitly addresses the 2208 issue and describes how it was solved.

The old problem was not the drivers but the configurations of the drivers, Creality uses the drivers in Standalone mode and LA was incompatible with that.

https://github.com/mriscoc/Ender3V2S1/blob/1c733afd3d533959f6822341fadfc77dc88a9e72/Marlin/Configuration.h#L179

Some months ago, that was resolved by this: https://github.com/MarlinFirmware/Marlin/pull/24533

**Reply by DanMan32** [reply]
Almost missed that last link.  You might want to include that in the general info about LA.
I am certainly going to try it.

**Reply by Triffen** [reply]
Do try it, it works fine on my 4.2.2 board.
